clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -162,6 +188,19 @@ public synchronized void update(Cluster cluster, long now) {
         this.lastSuccessfulRefreshMs = now;
         this.version += 1;
 
+        // Handle expiry of topics from the metadata list
+        Iterator<Map.Entry<String, Long>> it = topics.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<String, Long> entry = it.next();
+            Long expireMs = entry.getValue();
+            if (expireMs == TOPIC_EXPIRY_NEEDS_UPDATE)
+                entry.setValue(now + metadataExpireMs);
```

Comment:
I'm not sure how significant it is for the timeout to be a multiple of the refresh interval. The scheduling might not ever align anyway since it depends on poll() getting invoked at the right time. I also don't see why a separate mechanism would be needed for a hard-coded value. We're not expecting high granularity, just a way to avoid the cache growing unbounded over time. My concern is that we are technically changing the semantics of `metadata.max.age.ms` for the producer. Before it only controls how long we wait before refreshing metadata; now it also sets an expectation on the frequency of writes to each topic. Admittedly, the change should be transparent to the user, but it feels like an unneeded dependence.